Add missing IANA registrations (#144)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

          <author fullname="Vladimir Dzhuvinov">
            <organization>Connect2id</organization>
          </author>
          <date day="4" month="December" year="2023"/>
        </front>
</reference>

<reference anchor="IANA.OAuth.Parameters" target="https://www.iana.org/assignments/oauth-parameters">
  <front>
    <title>OAuth Parameters</title>
    <author>
      <organization>IANA</organization>
    </author>
    <date/>
  </front>
</reference>

<reference anchor="IANA.MediaTypes" target="https://www.iana.org/assignments/media-types">
  <front>
    <title>Media Types</title>
    <author>
      <organization>IANA</organization>
    </author>
    <date/>
  </front>
</reference>

# IANA Considerations

## Sub-Namespace Registration

This specification registers the following URN in the IANA "OAuth URI" registry [@!IANA.OAuth.Parameters] established by [@!RFC6755].

* URN: urn:ietf:params:oauth:grant-type:pre-authorized_code
* Common Name: Pre-Authorized Code
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#token_request) of this spedification

> ## OAuth Parameters Registry

Reviewer: I think this section also needs to register the following parameters:
* pre-authorized_code
* tx_code
* user_hint
* wallet_issuer (or whatever gets agreed upon in PR #142 )

This specification registers the following parameter names in the IANA "OAuth Parameters" registry [@!IANA.OAuth.Parameters] established by [@!RFC6749].

> 

Reviewer: ```suggestion

* Parameter Name: wallet_issuer
* Parameter Usage Location: authorization request
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#credential-authz-request) of this specification

* Parameter Name: user_hint
* Parameter Usage Location: authorization request
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#credential-authz-request) of this specification

```

* Parameter Name: issuer_state
* Parameter Usage Location: authorization request
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#credential-authz-request) of this specification

> 

Reviewer: ```suggestion

* Parameter Name: pre-authorized_code
* Parameter Usage Location: token request
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#token_request) of this specification

* Parameter Name: tx_code
* Parameter Usage Location: token request
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#token_request) of this specification

```

* Parameter Name: c_nonce
* Parameter Usage Location: token response
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#token-response) of this specification

* Parameter Name: c_nonce_expires_in
* Parameter Usage Location: token response
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#token-response) of this specification

## OAuth Dynamic Client Registration Metadata Registry

This specification registers the following client metadata name in the IANA "OAuth Dynamic Client Registration Metadata" registry [@!IANA.OAuth.Parameters] established by [@!RFC7591].

* Client Metadata Name: credential_offer_endpoint
* Client Metadata Description: Credential Offer Endpoint
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net

> * Reference: (#credential_offer_endpoint) of this specification

Reviewer: there is a also a need for a section `# OAuth Extensions Error Registry` that registers the following error names with "resource access error response" usage location:
* invalid_credential_request
* unsupported_credential_type
* unsupported_credential_format
* invalid_proof
* invalid_encryption_parameters

Reviewer: ```suggestion
* Reference: (#credential_offer_endpoint) of this specification

## OAuth Extensions Error Registry

This specification registers the following error values in the IANA "OAuth Extensions Error Registry" [@!IANA.OAuth.Params] established by [@!RFC6749].

* Name: invalid_credential_request
* Usage Location: resource access error response
* Protocol Extension: OpenID for Verifiable Credentials Issuance
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#credential-request-error) of this specification

* Name: unsupported_credential_type
* Usage Location: resource access error response
* Protocol Extension: OpenID for Verifiable Credentials Issuance
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#credential-request-error) of this specification

* Name: unsupported_credential_format
* Usage Location: resource access error response
* Protocol Extension: OpenID for Verifiable Credentials Issuance
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#credential-request-error) of this specification

* Name: invalid_proof
* Usage Location: resource access error response
* Protocol Extension: OpenID for Verifiable Credentials Issuance
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#credential-request-error) of this specification

* Name: invalid_encryption_parameters
* Usage Location: resource access error response
* Protocol Extension: OpenID for Verifiable Credentials Issuance
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#credential-request-error) of this specification
```

* Reference: (#credential-authz-request) of this specification

* Parameter Name: pre-authorized_code
* Parameter Usage Location: token request
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#token_request) of this specification

* Parameter Name: tx_code
* Parameter Usage Location: token request
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#token_request) of this specification

* Parameter Name: c_nonce
* Parameter Usage Location: token response
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#token-response) of this specification

* Parameter Name: c_nonce_expires_in
* Parameter Usage Location: token response
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net

> * Reference: (#token-response) of this specification

Reviewer 2: I would prefer to see these removed (a la https://github.com/openid/OpenID4VCI/issues/39) but I recognize this is in vain and am just putting something here for posterity.

## OAuth Dynamic Client Registration Metadata Registry

This specification registers the following client metadata name in the IANA "OAuth Dynamic Client Registration Metadata" registry [@!IANA.OAuth.Parameters] established by [@!RFC7591].

* Client Metadata Name: credential_offer_endpoint
* Client Metadata Description: Credential Offer Endpoint
* Change Controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#credential_offer_endpoint) of this specification

## OAuth Extensions Error Registry

This specification registers the following error values in the IANA "OAuth Extensions Error Registry" [@!IANA.OAuth.Params] established by [@!RFC6749].

* Name: invalid_credential_request
* Usage Location: resource access error response
* Protocol Extension: OpenID for Verifiable Credentials Issuance
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net

> * Reference: (#credential-request-error) of this specification

Reviewer: ```suggestion
* Reference: (#credential-request-errors) of this specification
```

* Name: unsupported_credential_type
* Usage Location: resource access error response
* Protocol Extension: OpenID for Verifiable Credentials Issuance
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net

> * Reference: (#credential-request-error) of this specification

Reviewer: ```suggestion
* Reference: (#credential-request-errors) of this specification
```

* Name: unsupported_credential_format
* Usage Location: resource access error response
* Protocol Extension: OpenID for Verifiable Credentials Issuance
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Reference: (#credential-request-error) of this specification

* Name: invalid_proof
* Usage Location: resource access error response
* Protocol Extension: OpenID for Verifiable Credentials Issuance
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net

> * Reference: (#credential-request-error) of this specification

Reviewer: ```suggestion
* Reference: (#credential-request-errors) of this specification
```

* Name: invalid_encryption_parameters
* Usage Location: resource access error response
* Protocol Extension: OpenID for Verifiable Credentials Issuance
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net

> * Reference: (#credential-request-error) of this specification

Reviewer: ```suggestion
* Reference: (#credential-request-errors) of this specification
```

> * Reference: (#credential-request-errors) of this specification

Reviewer 2: These are not OAuth errors and therefore don't need to be registered and shouldn't be.  see https://datatracker.ietf.org/doc/html/rfc6749#section-7.2

Author: These are errors defined by a profile of OAuth.  Other errors defined by other OAuth profiles *are* registered.  For instance, https://openid.net/specs/openid-connect-core-1_0.html#ErrorContents registered errors defined by the OAuth profile called OpenID Connect.  You can see their registrations at https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#extensions-error.

These errors defined by the OAuth profile called OpenID for Verifiable Credential Issuance also should and need to be registered.

Reviewer 2: > These are errors defined by a profile of OAuth. Other errors defined by other OAuth profiles _are_ registered. For instance, https://openid.net/specs/openid-connect-core-1_0.html#ErrorContents registered errors defined by the OAuth profile called OpenID Connect. You can see their registrations at https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#extensions-error.

Yes but all of those registered OpenID Connect errors are returned from the authorization endpoint using the standard parameter for such.  Registration for those makes sense and is one of those prescribed by https://datatracker.ietf.org/doc/html/rfc6749#section-8.5 (note that general protected resource application errors are not). 

> These errors defined by the OAuth profile called OpenID for Verifiable Credential Issuance also should and need to be registered.

These are errors for the Credential Request endpoint, which is an OAuth protected resource but is not an OAuth endpoint. The only protected resource errors that need registration are "for error values to be shared among OAuth token authentication schemes" (such as Bearer / DPoP) for when a "resource **access** request fails"  https://datatracker.ietf.org/doc/html/rfc6749#section-7.2 

The registry is not for general application layer errors from the protected resource.

> ## OAuth Extensions Error Registry
> 
> This specification registers the following error values in the IANA "OAuth Extensions Error Registry" [@!IANA.OAuth.Parameters] established by [@!RFC6749].
> 
> * Name: invalid_credential_request
> * Usage Location: resource access error response
> * Protocol Extension: OpenID for Verifiable Credentials Issuance
> * Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
> * Reference: (#credential-request-errors) of this specification
> 
> * Name: unsupported_credential_type
> * Usage Location: resource access error response
> * Protocol Extension: OpenID for Verifiable Credentials Issuance
> * Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
> * Reference: (#credential-request-errors) of this specification
> 
> * Name: unsupported_credential_format
> * Usage Location: resource access error response
> * Protocol Extension: OpenID for Verifiable Credentials Issuance
> * Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
> * Reference: (#credential-request-errors) of this specification
> 
> * Name: invalid_proof
> * Usage Location: resource access error response
> * Protocol Extension: OpenID for Verifiable Credentials Issuance
> * Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
> * Reference: (#credential-request-errors) of this specification
> 
> * Name: invalid_encryption_parameters
> * Usage Location: resource access error response
> * Protocol Extension: OpenID for Verifiable Credentials Issuance
> * Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
> * Reference: (#credential-request-errors) of this specification

Reviewer 2: ```suggestion
```

This specification registers the following well-known URI in the IANA "Well-Known URI" registry established by [@!RFC5785].

* URI suffix: openid-credential-issuer
* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Specification document: (#credential-issuer-wellknown) of this document
* Related information: (none)

## Media Types Registry

This specification registers the following media types in the IANA "Media Types" registry [@!IANA.MediaTypes] in the manner described in [@!RFC6838].

* Type name: `application`
* Subtype name: `openid4vci-proof+jwt`
* Required parameters: n/a
* Optional parameters: n/a
* Encoding considerations: Uses JWS Compact Serialization, as specified in [@!RFC7515].
* Security considerations: See the Security Considerations in [@!RFC7519].
* Interoperability considerations: n/a
* Published specification: (#jwt-proof-type) of this specification
* Applications that use this media type: Applications that issue and store verifiable credentials
* Additional information:
  - Magic number(s): n/a
  - File extension(s): n/a
  - Macintosh file type code(s): n/a

> * Person & email address to contact for further information: TBD

Reviewer 3: TBD? Why not the draft authors or the WG?

Reviewer: +1 
```suggestion
* Person & email address to contact for further information: Torsten Lodderstedt, torsten@lodderstedt.net
```

* Intended usage: COMMON
* Restrictions on usage: none

> * Author: Michael B. Jones, michael_b_jones@hotmail.com

Reviewer 3: Why does this registration require an author? if so, it should be the draft authors, shouldn't it?

Author: An author is required because it's required in the registration template at https://www.rfc-editor.org/rfc/rfc6838.html#section-5.6.  The odd thing about this template is that the Author is separate from the Change Controller.  (For most IANA registries there's no Author - only a Change Controller.)

I used my name since I'm the author of the registration but I don't feel strongly about it staying that way.  One thing we could try is making the Author value the same as the Change Controller value and see if that flies.  What do you think?

Reviewer: Agree with Torsten, this should be one of the spec authors

```suggestion
* Author: Torsten Lodderstedt, torsten@lodderstedt.net
```

* Change controller: OpenID Foundation Digital Credentials Protocols Working Group - openid-specs-digital-credentials-protocols@lists.openid.net
* Provisional registration? No

* Type name: `application`
* Subtype name: `openid4vci-proof+cwt`
* Required parameters: n/a
* Optional parameters: n/a
* Encoding considerations: Binary CBOR, as specified in [@!RFC9052]
* Security considerations: See the Security Considerations in [@!RFC8392].
* Interoperability considerations: n/a
* Published specification: (#cwt-proof-type) of this specification
* Applications that use this media type: Applications that issue and store verifiable credentials
* Additional information:
  - Magic number(s): n/a
  - File extension(s): n/a
  - Macintosh file type code(s): n/a

> * Person & email address to contact for further information: TBD

Reviewer 3: same as above

Reviewer: ```suggestion
* Person & email address to contact for further information: Torsten Lodderstedt, torsten@lodderstedt.net
```

* Intended usage: COMMON
* Restrictions on usage: none

> * Author: Michael B. Jones, michael_b_jones@hotmail.com

Reviewer 3: same as above

Reviewer: ```suggestion
* Author: Torsten Lodderstedt, torsten@lodderstedt.net
```
